blackjax/mcmc/integrators.py

```diff
@@ -184,65 +271,108 @@ def yoshida(
     a1 = 0.29619504261126
     b2 = 0.5 - b1
     a2 = 1 - 2 * a1
+    cofficients = [b1, a1, b2, a2, b2, a1, b1]
+    position_update_fn = euclidean_position_update_fn(logdensity_fn)
+    momentum_update_fn = euclidean_momentum_update_fn(kinetic_energy_fn)
+    one_step = generalized_symplectic_integrator(
+        momentum_update_fn,
+        position_update_fn,
+        cofficients,
+        format_output_fn=format_euclidean_state_output,
+    )
 
-    logdensity_and_grad_fn = jax.value_and_grad(logdensity_fn)
-    kinetic_energy_grad_fn = jax.grad(kinetic_energy_fn)
-
-    def one_step(state: IntegratorState, step_size: float) -> IntegratorState:
-        position, momentum, _, logdensity_grad = state
-
-        momentum = jax.tree_util.tree_map(
-            lambda momentum, logdensity_grad: momentum
-            + b1 * step_size * logdensity_grad,
-            momentum,
-            logdensity_grad,
-        )
-
-        kinetic_grad = kinetic_energy_grad_fn(momentum)
-        position = jax.tree_util.tree_map(
-            lambda position, kinetic_grad: position + a1 * step_size * kinetic_grad,
-            position,
-            kinetic_grad,
-        )
-
-        _, logdensity_grad = logdensity_and_grad_fn(position)
-        momentum = jax.tree_util.tree_map(
-            lambda momentum, logdensity_grad: momentum
-            + b2 * step_size * logdensity_grad,
-            momentum,
-            logdensity_grad,
-        )
+    return one_step
 
-        kinetic_grad = kinetic_energy_grad_fn(momentum)
-        position = jax.tree_util.tree_map(
-            lambda position, kinetic_grad: position + a2 * step_size * kinetic_grad,
-            position,
-            kinetic_grad,
-        )
 
-        _, logdensity_grad = logdensity_and_grad_fn(position)
-        momentum = jax.tree_util.tree_map(
-            lambda momentum, logdensity_grad: momentum
-            + b2 * step_size * logdensity_grad,
-            momentum,
-            logdensity_grad,
-        )
+# Intergrators with non Euclidean updates
+def esh_dynamics_momentum_update_one_step(
```

Comment:
@reubenharry could you check the implementation to see if it is correct? I ended up using `ravel_pytree(logdensity_grad)` because otherwise using the `tree_map` and `tree_reduce` gets complicated (also complicated for dot product).